### ISSUE TITLE
Failing to Replicate Main Results with Llama3.2-1B-instruct

### ISSUE THREAD

**Issue by X1AOX1A** [yc1999/LLM2#1]
I attempted to replicate the main results using the provided train_verifier.sh and evaluate_gsm8k.sh scripts on a setup with 4xA800-80G GPUs. However, the results are significantly lower than expected, with only 2% accuracy achieved on the GSM8K dataset.

Is there a potential issue with the configuration, scripts, or my setup? Any guidance would be appreciated.

**Results**

> As the `data/evaluation/gsm8k/test.jsonl` is missing, I replaced with the [openai/gsm8k test set](https://huggingface.co/datasets/openai/gsm8k/viewer/main/test) with 1319 samples.

```
{
    "num": 1319,
    "accuracy": 0.02350265352539803,
    "time": "309.70 seconds",
    "average_generation_time": 0.827823323914499
}
```

**Training Loss of verifer**
<img width="1474" alt="Screenshot 2025-01-01 at 23 10 08" src="https://github.com/user-attachments/assets/7afc3f1a-2f9b-47d3-b889-9ff29d2fb6c8" />

**train_verifier.sh**

```bash
MODEL_NAME_OR_PATH=meta-llama/Llama-3.2-1B-Instruct

OUTPUT_DIR=./checkpoints/${MODEL}-Verifier
mkdir -p $OUTPUT_DIR

NUM_PROCESSES=4
TOTAL_BATCH_SIZE=128
BATCH_SIZE_PER_GPU=16
GRADIENT_ACC_STEPS=$(($TOTAL_BATCH_SIZE/$NUM_PROCESSES/$BATCH_SIZE_PER_GPU))

ACCELERATE_LOG_LEVEL=info accelerate launch \
        --main_process_port 39999 \
        --num_processes $NUM_PROCESSES \
        --config_file config_files/accelerate_ds_zero3.yaml \
        train_verifier.py \
        config_files/verifier.yaml \
        --model_name_or_path=$MODEL_NAME_OR_PATH \
        --per_device_train_batch_size=$BATCH_SIZE_PER_GPU \
        --per_device_eval_batch_size=$BATCH_SIZE_PER_GPU \
        --gradient_accumulation_steps=$GRADIENT_ACC_STEPS
```

**evaluate_gsm8k.sh**

```bash
# export NCCL_P2P_LEVEL=NVL
# export NCCL_IB_GID_INDEX=3
export HF_DATASETS_OFFLINE=1
export CUDA_VISIBLE_DEVICES=0,1,2,3

benchmark=gsm8k
n_shot=0
prompt_format=llama
max_new_tokens=512
max_input_tokens=2048
micro_batch_size_per_gpu=1

MODEL=Llama-3.2-1B-Instruct
MODEL_NAME_OR_PATH=meta-llama/Llama-3.2-1B-Instruct

VERIFIER=${MODEL}-Verifier
VERIFIER_MODEL_NAME_OR_PATH=./checkpoints/${VERIFIER}

alpha=0.1
beta=0.25
num_beams=1


for beta in 0.25
do

args="
    --benchmark ${benchmark} \
    --data_dir ./data/evaluation/${benchmark} \
    --n_shot ${n_shot} \
    --max_new_tokens ${max_new_tokens} \
    --max_input_tokens ${max_input_tokens} \
    --micro_batch_size_per_gpu ${micro_batch_size_per_gpu} \
    --model_name_or_path ${MODEL_NAME_OR_PATH} \
    --num_workers 4 \
    --preprocessing_num_workers 1 \
    --keep_in_memory \
    --num_beams ${num_beams} \
"

if [ "$prompt_format" != "None" ]; then
    args="${args} --prompt_format ${prompt_format}"
fi

if [ "$VERIFIER" == "None" ]; then
    args="
    ${args} \
    --save_dir ./results/${benchmark}/${MODEL}/${n_shot}-${prompt_format}-"${prefix}"-${max_new_tokens}-${max_input_tokens}-${micro_batch_size_per_gpu}-${num_beams}
"
else
    args="
    ${args} \
    --verifier_model_name_or_path ${VERIFIER_MODEL_NAME_OR_PATH} \
    --alpha ${alpha} \
    --beta ${beta} \
    --save_dir ./results/${benchmark}/${MODEL}/${VERIFIER}/${n_shot}-${prompt_format}-"${prefix}"-${max_new_tokens}-${max_input_tokens}-${micro_batch_size_per_gpu}-${strategy}-${alpha}-${beta}-${top_k}-${num_beams}
"
fi

accelerate launch evaluation.py $args

done
```

**Environment** (installed with requirements.txt and python3.10):

```
accelerate               0.34.2
aiohappyeyeballs         2.4.4
aiohttp                  3.11.11
aiosignal                1.3.2
annotated-types          0.7.0
anyio                    4.7.0
appdirs                  1.4.4
async-timeout            4.0.3
attrs                    24.3.0
bitarray                 2.9.2
bitsandbytes             0.42.0
blessed                  1.20.0
certifi                  2024.12.14
charset-normalizer       3.4.1
click                    8.1.8
dataclasses-json         0.6.7
datasets                 3.0.1
deepspeed                0.15.1
dill                     0.3.8
docker-pycreds           0.4.0
docstring_parser         0.16
evaluate                 0.4.1
exceptiongroup           1.2.2
filelock                 3.16.1
frozenlist               1.5.0
fsspec                   2024.6.1
gitdb                    4.0.11
GitPython                3.1.43
gpustat                  1.1.1
greenlet                 3.1.1
hjson                    3.1.0
huggingface-hub          0.24.7
idna                     3.10
Jinja2                   3.1.5
jsonpatch                1.33
jsonpointer              3.0.0
langchain                0.1.0
langchain-community      0.0.20
langchain-core           0.1.23
langsmith                0.0.87
markdown-it-py           3.0.0
MarkupSafe               3.0.2
marshmallow              3.23.2
mdurl                    0.1.2
mpmath                   1.3.0
multidict                6.1.0
multiprocess             0.70.16
mypy-extensions          1.0.0
networkx                 3.4.2
ninja                    1.11.1.1
numpy                    1.26.4
nvidia-cublas-cu12       12.4.5.8
nvidia-cuda-cupti-cu12   12.4.127
nvidia-cuda-nvrtc-cu12   12.4.127
nvidia-cuda-runtime-cu12 12.4.127
nvidia-cudnn-cu12        9.1.0.70
nvidia-cufft-cu12        11.2.1.3
nvidia-curand-cu12       10.3.5.147
nvidia-cusolver-cu12     11.6.1.9
nvidia-cusparse-cu12     12.3.1.170
nvidia-ml-py             12.560.30
nvidia-nccl-cu12         2.21.5
nvidia-nvjitlink-cu12    12.4.127
nvidia-nvtx-cu12         12.4.127
openai                   0.28.0
packaging                23.2
pandas                   2.2.3
peft                     0.13.0
pip                      24.2
propcache                0.2.1
protobuf                 4.25.5
psutil                   6.1.1
py-cpuinfo               9.0.0
pyarrow                  18.1.0
pydantic                 2.10.4
pydantic_core            2.27.2
Pygments                 2.18.0
python-dateutil          2.9.0.post0
pytz                     2024.2
PyYAML                   6.0.2
regex                    2024.11.6
requests                 2.32.3
responses                0.18.0
rich                     13.9.4
safetensors              0.4.5
scipy                    1.14.1
sentencepiece            0.2.0
sentry-sdk               2.19.2
setproctitle             1.3.4
setuptools               75.1.0
shortuuid                1.0.13
shtab                    1.7.1
six                      1.17.0
smmap                    5.0.1
sniffio                  1.3.1
SQLAlchemy               2.0.36
sympy                    1.13.1
tenacity                 8.5.0
tensorboardX             2.6.2.2
tokenizers               0.20.3
torch                    2.5.1
tqdm                     4.67.1
transformers             4.45.1
triton                   3.1.0
trl                      0.8.3
typeguard                4.4.1
typing_extensions        4.12.2
typing-inspect           0.9.0
tyro                     0.9.5
tzdata                   2024.2
urllib3                  2.3.0
wandb                    0.16.2
wcwidth                  0.2.13
wheel                    0.44.0
xxhash                   3.5.0
yarl                     1.18.3
```
